python/utils/OpaqueArguments.h

```diff
@@ -156,212 +163,197 @@ inline py::args simplifiedValidateInputArguments(py::args &args) {
+                                              py::object &arg)> &backupHandler,
+                     bool checkNumArgs = false) {
+  if (checkNumArgs)
+    if (kernelFuncOp.getNumArguments() != args.size())
+      throw std::runtime_error("Invalid runtime arguments - kernel expected " +
+                               std::to_string(kernelFuncOp.getNumArguments()) +
+                               " but was provided " +
+                               std::to_string(args.size()) + " arguments.");
 
-inline void
-packArgs(OpaqueArguments &argData, py::args args,
-         const std::function<bool(OpaqueArguments &argData, py::object &arg)>
-             &backupHandler) {
   for (std::size_t i = 0; i < args.size(); i++) {
     py::object arg = args[i];
-    if (py::isinstance<py::float_>(arg)) {
-      double *ourAllocatedArg = new double();
-      *ourAllocatedArg = PyFloat_AsDouble(arg.ptr());
-      argData.emplace_back(ourAllocatedArg, [](void *ptr) {
-        delete static_cast<double *>(ptr);
-      });
-      continue;
-    }
-
-    if (py::isinstance<py::int_>(arg)) {
-      long *ourAllocatedArg = new long();
-      *ourAllocatedArg = PyLong_AsLong(arg.ptr());
-      argData.emplace_back(ourAllocatedArg,
-                           [](void *ptr) { delete static_cast<long *>(ptr); });
-      continue;
-    }
-
-    if (py::isinstance<cudaq::pauli_word>(arg)) {
-      cudaq::pauli_word *ourAllocatedArg =
-          new cudaq::pauli_word(arg.cast<cudaq::pauli_word>().str());
-      argData.emplace_back(ourAllocatedArg, [](void *ptr) {
-        delete static_cast<cudaq::pauli_word *>(ptr);
-      });
-      continue;
-    }
-
-    if (py::isinstance<py::list>(arg)) {
-      auto casted = py::cast<py::list>(arg);
-      if (casted.empty()) {
-        // If its empty, just put any vector on the `argData`,
-        // it won't matter since it is empty and all
-        // vectors have the same memory footprint (span-like).
-        std::vector<bool> *ourAllocatedArg = new std::vector<bool>();
-        argData.emplace_back(ourAllocatedArg, [](void *ptr) {
-          delete static_cast<std::vector<bool> *>(ptr);
-        });
-        continue;
-      }
-
-      // Get the first element in the list
-      auto firstElement = casted[0];
-
-      // Handle `list[pauli_word]`
-      if (py::isinstance<cudaq::pauli_word>(firstElement)) {
-        std::vector<cudaq::pauli_word> *ourAllocatedArg =
-            new std::vector<cudaq::pauli_word>(casted.size());
-        for (std::size_t counter = 0; auto el : casted) {
-          auto pw = el.cast<cudaq::pauli_word>();
-          (*ourAllocatedArg)[counter++] = cudaq::pauli_word(pw.str());
-        }
-        argData.emplace_back(ourAllocatedArg, [](void *ptr) {
-          delete static_cast<std::vector<cudaq::pauli_word> *>(ptr);
+    auto kernelArgTy = kernelFuncOp.getArgument(i).getType();
+    llvm::TypeSwitch<mlir::Type, void>(kernelArgTy)
```

Comment:
I'm curious - does `TypeSwitch` construct all objects regardless of which `Case` you enter?

Reply:
Eric may be the best to answer this, but I think that TypeSwitch constructs a visitor pattern under the hood and requires construction of only the lambda type overloads that you provide, while all other types default to the `Default()` method lambda you provide.